utils/uri/matcher.go

```diff
@@ -0,0 +1,75 @@
+package uri
+
+import (
+	"errors"
+	"strings"
+)
+
+var (
+	ErrInvalidURI = errors.New("invalid URI string")
+)
+
+//hs:<service_name>:<subsystem>:<version>:<category>:<resource_name>?<allowed_arguments>#<permission_metadata>
+type URI struct {
+	path      string
+	arguments map[string]string
+	metadata  map[string]string
+}
+
+// hs:<service_name>:<subsystem>:<version>:<category>:<resource_name>?<allowed_arguments>#<permission_metadata>
+func ParseURI(source string) (*URI, error) {
```

Comment:
Maybe we should just return an object here instead of a pointer? Might make more sense if we want the URIs to be immutable.

Also, how would you feel about renaming this to `NewURIFromString`? [I used this pattern in my PR for resource -> URI mapping](https://github.com/unicsmcr/hs_auth/pull/76/files#diff-65fb66b2882725dc11d1f17db236fa9eR20) and it might be a bit cleaner because we'll need to support multiple ways to construct a URI object.

Reply:
Yep, thanks for pointing that out, I'll make sure the naming matches 🙂